File: frontend/src/types/skill.ts
/**
 * Skill types
 */

export interface ModelOverrides {
  temperature?: number;
  maxOutputTokens?: number;
  thinkingMode?: boolean;
  model?: string;
}

export interface Skill {
  id: string;
  name: string;
  description: string;
  prompt: string;
  color: string;
  tokenCount: number;
  isBuiltIn: boolean;
  createdAt: number;
  lastUsedAt: number;
  // Enhanced skill dimensions
  toolIds?: string[];
  files?: string[];
  contextIds?: string[];
  modelOverrides?: ModelOverrides;
  // Discovery metadata
  source?: 'builtin' | 'custom' | 'project' | 'user';
  allowImplicitInvocation?: boolean;
  // agentskills.io spec fields
  keywords?: string[];
  license?: string;
  compatibility?: string;
  skillMetadata?: Record<string, string>;
  allowedTools?: string[];
  skillPath?: string;
  hasScripts?: boolean;
  hasReferences?: boolean;
  hasAssets?: boolean;
}

/** True if this skill carries more than just a prompt. */
export function isEnhancedSkill(skill: Skill): boolean {
  return !!(
    (skill.toolIds && skill.toolIds.length > 0) ||
    (skill.files && skill.files.length > 0) ||
    (skill.contextIds && skill.contextIds.length > 0) ||
    skill.modelOverrides
  );
}

export interface SkillCreate {
  name: string;
  description: string;
  prompt: string;
  toolIds?: string[];
  files?: string[];
  contextIds?: string[];
  modelOverrides?: ModelOverrides;
  allowImplicitInvocation?: boolean;
}

export interface SkillUpdate {
  name?: string;
  description?: string;
  prompt?: string;
  toolIds?: string[];
  files?: string[];
  contextIds?: string[];
  modelOverrides?: ModelOverrides;
  allowImplicitInvocation?: boolean;
}

File: app/models/skill.py
"""
Skill data models.
"""
from pydantic import BaseModel
from typing import Optional, List, Dict, Any


class ModelOverrides(BaseModel):
    """Optional model parameter overrides applied when a skill is active."""
    model_config = {"extra": "allow"}

    temperature: Optional[float] = None
    maxOutputTokens: Optional[int] = None
    thinkingMode: Optional[bool] = None
    model: Optional[str] = None


class Skill(BaseModel):
    id: str
    name: str
    description: str
    prompt: str
    color: str
    tokenCount: int
    isBuiltIn: bool = False
    createdAt: int
    lastUsedAt: int
    # Enhanced skill dimensions (all optional for backward compat)
    toolIds: Optional[List[str]] = None
    files: Optional[List[str]] = None
    contextIds: Optional[List[str]] = None
    modelOverrides: Optional[ModelOverrides] = None
    # Discovery metadata
    source: Optional[str] = None  # 'builtin', 'custom', 'project', 'user'
    allowImplicitInvocation: bool = True
    # agentskills.io spec fields
    keywords: Optional[List[str]] = None
    license: Optional[str] = None
    compatibility: Optional[str] = None
    skillMetadata: Optional[Dict[str, str]] = None
    allowedTools: Optional[List[str]] = None
    skillPath: Optional[str] = None  # Filesystem path for project-discovered skills
    hasScripts: bool = False
    hasReferences: bool = False
    hasAssets: bool = False


class SkillCreate(BaseModel):
    name: str
    description: str
    prompt: str

    toolIds: Optional[List[str]] = None
    files: Optional[List[str]] = None
    contextIds: Optional[List[str]] = None
    modelOverrides: Optional[ModelOverrides] = None
    allowImplicitInvocation: Optional[bool] = None


class SkillUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    prompt: Optional[str] = None

    toolIds: Optional[List[str]] = None
    files: Optional[List[str]] = None
    contextIds: Optional[List[str]] = None
    modelOverrides: Optional[ModelOverrides] = None
    allowImplicitInvocation: Optional[bool] = None
